pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -569,42 +292,50 @@ func (c *Cluster) startOSDDaemonsOnNode(nodeName string, config *provisionConfig
-			continue
-		}
+	if err := createOrUpdateOSDDeployment(c, osd, osdProps, osdLongName, config); err != nil {
+		return err
+	}
 
-		// Set the deployment hash as an annotation
-		err = patch.DefaultAnnotator.SetLastAppliedAnnotation(dp)
-		if err != nil {
-			errMsg := fmt.Sprintf("failed to set annotation for deployment %q. %v", dp.Name, err)
-			config.addError(errMsg)
-			continue
-		}
+	return nil
+}
 
-		_, createErr := c.context.Clientset.AppsV1().Deployments(c.clusterInfo.Namespace).Create(ctx, dp, metav1.CreateOptions{})
-		if createErr != nil {
-			if kerrors.IsAlreadyExists(createErr) {
-				logger.Debugf("deployment for osd %d already exists. updating if needed", osd.ID)
-				if err = updateDeploymentAndWait(c.context, c.clusterInfo, dp, opconfig.OsdType, strconv.Itoa(osd.ID), c.spec.SkipUpgradeChecks, c.spec.ContinueUpgradeAfterChecksEvenIfNotHealthy); err != nil {
-					logger.Errorf("failed to update osd deployment %d. %v", osd.ID, err)
-				}
-			} else {
-				// we failed to create job, update the orchestration status for this pvc
-				logger.Warningf("failed to create osd deployment for node %q, osd %+v. %v", n.Name, osd, createErr)
-				continue
+// osdLongName would be, for example, "OSD 1 on node k8s-worker-2" or "OSD 5 on PVC pvc-name"
+func createOrUpdateOSDDeployment(c *Cluster, osd OSDInfo, osdProps osdProperties, osdLongName string, config *provisionConfig) error {
+	dp, err := c.makeDeployment(osdProps, osd, config)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create deployment for %s", osdLongName)
+	}
+
+	// Set the deployment hash as an annotation
+	err = patch.DefaultAnnotator.SetLastAppliedAnnotation(dp)
+	if err != nil {
+		return errors.Wrapf(err, "failed to set last-applied annotation for on deployment %q for %s", dp.Name, osdLongName)
+	}
+
+	_, createErr := c.context.Clientset.AppsV1().Deployments(c.clusterInfo.Namespace).Create(context.TODO(), dp, metav1.CreateOptions{})
+	if createErr != nil {
+		if kerrors.IsAlreadyExists(createErr) {
+			if err = updateDeploymentAndWait(c.context, c.clusterInfo, dp, opconfig.OsdType, strconv.Itoa(osd.ID), c.spec.SkipUpgradeChecks, c.spec.ContinueUpgradeAfterChecksEvenIfNotHealthy); err != nil {
+				// TODO: should this register as a deployment failure? I think yes.
+				// logger.Errorf("failed to update osd deployment %d. %v", osd.ID, err)
+				return errors.Wrapf(err, "failed to update deployment for %s", osdLongName)
 			}
+			logger.Infof("updated deployment for %s", osdLongName)
 		} else {
-			logger.Infof("created deployment for osd %d", osd.ID)
+			// TODO: should this register as a deployment failure? I think yes.
+			// we failed to create job, update the orchestration status for this pvc
+			// logger.Warningf("failed to create osd deployment for node %q, osd %+v. %v", n.Name, osd, createErr)
+			return errors.Wrapf(err, "failed to create deployment for %s", osdLongName)
 		}
+	} else {
+		logger.Infof("created deployment for %s", osdLongName)
 	}
```

Comment:
@leseb @travisn (two TODO questions that are basically the same)

Reply:
Agreed, if the OSD fails to be created or updated it should fail the reconcile (after trying all the other OSDs). It will be added to the errors collection from the caller, right?

Reply:
Yes. In `create.go` `createNewOSDsFromStatus` adds errors returned from here (and the parent caller) to the collection of errors. Since this is a single-purpose, low-level method, I let the calling function decide how to handle the errors.